README.md
## What is it

Mathematical expression evaluation library with big integers, floats, common fractions, and complex numbers support.
The library is used by the project [RionaCalc](https://github.com/VladimirMarkelov/rclc)

## Features

* Automatic selection of more appropriate argument type for a function: e.g, `sqrt(-4)` converts float number `-4` into complex one `-4+0i` and then calculates the result `0+2i`. The same is true for calculating logarithm for negative float numbers, and acos and asin for argument greater than `1.0`
* Automatic adding multiplication sign where it is omitted: e.g, `(1+2)(2+9)` is calculated as `(1+2)*(2+9)`
* Functions with a single-value argument do not require to enclose its argument into brackets: e.g, `sin cos 2` is calculated as `sin(cos(2))`
* The final closing brackets can be omitted: e.g, `(1+2)*(2+9` is the same as `(1+2)*(2+9)`
* Trigonometric functions work with radians and degrees. Bare numbers are treated as radians, degrees requires one or three suffixes. Two degrees formats: `20d30m50s` or `20°30'50"`. Minutes and seconds can be omitted, in this case degrees can be float number like `30.25d`. So, `sin(pi/2)` == `sin(90°)`
* Every number can include group separator `_` for readability - it is very useful when using big integers. `3_000.90_23` == `3000.9023`
* Both `.` and `,` are treated as decimal separators
* Function argument separator is `;`. If a function receives more arguments than it requires, the trailing arguments are dropped: e.g, `sqrt(11;12;13)` is the same as `sqrt(11)`
* Regular fractions use `\` to separate its parts. They can be written with integer part or only with numerator and denominator, e.g `1\1\10` == `11\10`
* Two complex numbers formats: with marker at the end or in the middle. E.g, `1+2i` == `1+i2`. In addition, `j` can be used instead of `i` - but the calculator outputs always with `i`
* Hexadecimal(starts with `0x`), octal(starts with `0o`), and binary(starts with `0b`) numbers
* Character `%` can be either a modulo or a percentage operator. It depends on the character position: if `%` is right before the expression end or before closing bracket or before another operator *and* previous operator is one of `+`, `-`, `*`, or `/`, the character is considered a percentage operator


changelog
rcalclib (1.0.0) unstable; urgency=medium

  * '**' and '^' are synonyms and both mean "raise to power", for bit-wise xor
    use '@'

  * Percentage operator priority lowered. Now it has the same priority as '+'
    and '-'

  * '%' works as percentage operator in some cases: when it is the last symbol
    in an expression or it is right before closing bracket or another operator
    'a + n %' = increase 'a' by 'n' percents
    'a - n %' = decrease 'a' by 'n' percents
    'a * n %' = calculate 'n' percents of 'a'
    'a / b %' = calculate how many percents is 'a' of 'b'

  * Clippy fixes and package upgrades

 -- Vladimir Markelov <vmatroskin@gmail.com>  Thu, 08 Jul 2021 18:55:24 -0700

rcalclib (0.9.3) unstable; urgency=medium

  * Sqrt for positive common fraction returns common fraction if possible

  * New function: solve(aliases: 'zeros' and 'roots') calculates roots of
    linear or quadratic eqaution. The result is the first root.

 -- Vladimir Markelov <vmatroskin@gmail.com>  Tue, 21 May 2019 19:41:51 -0700

rcalclib (0.9.2) unstable; urgency=medium

  * Fix raising negative float number into non-integer power degree

  * New function: gamma. The calculation is approximate, its accuracy for
    integer results that fit f64(argument < 20) is around +/- 0.000000001%

 -- Vladimir Markelov <vmatroskin@gmail.com>  Mon, 13 May 2019 20:22:51 -0700

rcalclib (0.9.1) unstable; urgency=medium

  * function ratio can create a common fraction from two integers: numerator
    and denominator. Example: ratio(2;5) = 2\5

  * functions min and max to return minimum and maximum value from the list of
    arbitrary length

  * functions avg to calculate average of a list of values

  * functions is_prime and next_prime: work only with integers and use naive
    algorithm that makes them slow for big numbers

  * functions hex, oct, and bin to print an integer in non-decimal format

 -- Vladimir Markelov <vmatroskin@gmail.com>  Sat, 04 May 2019 19:47:14 -0700

rcalclib (0.9.0) unstable; urgency=medium

  * New functions: fib(N-th Fibonacci number), cbrt(cubic root)

 -- Vladimir Markelov <vmatroskin@gmail.com>  Sat, 27 Apr 2019 20:19:29 -0700

rcalc (0.8.0) unstable; urgency=medium

  * First stable version

 -- Vladimir Markelov <vmatroskin@gmail.com>  Sat, 20 Apr 2019 23:33:19 -0700


src/lib.rs
//! # Expression calculator
//!
//! Calculator transparently supports operations with different type of numbers,
//! converting to appropriate type when it is needed. For instance,
//! `sqrt(-2)` is automatically converted from integer number `-2` to complex
//! one `-2+0i` and is evaluated to `0+1.4142135623730951i`
//!
//! Extra feature: trigonometric functions support both radians and degrees:
//! * Radians are just float numbers: `sin(pi/2)` returns `1`
//! * Degrees can be written in one of the following forms: `12°30'45"` or `12d30m45s` or `12.5125d`
//!
//! Internally degrees are stored and act as float numbers. They only support
//! their own input format. Output format is always a float number.
//!
//! If two numbers have different types, they are converted to highest type and
//! then the result is calculated. The list of number types starting from highest:
//! * Complex numbers. Two input formats supported: `1+2i` and `1+i2`. `i` can be capital. `j` can
//! be used instead of `i`
//! * Float numbers (degrees are float numbers, too)
//! * Rational numbers
//! * Big integer numbers
//!
//! Note: sometimes higher types are converted to lower ones. It may happen
//! after multiplication, division, and squaring a number. Examples:
//! * `sqr(0+2i)` -> `-4` - automatically complex is converted to float number
//! * `1.5 / 0.5` -> `3` - float to big integer
//! * `1\2 + 1\2` -> `1` - half and half is an integer one
//!
//! The list of supported functions:
//! * trigonometric functions (including inverted ones): sin, cos, tan, asin, acos, atan
//! * hyperbolic functions (including inverted ones): sinh, cosh, tanh, asinh, acosh, atanh
//! * square and square root: sqr and sqrt
//! * exponent, logarithm: exp, ln
//! * complex functions: norm, re, im, conjugate
//! * rounding: ceil, floor, trunc, round
//! * float to rational: ratio
//! * absolute value and sign: abs, signum
//! * fractional part of a float number: fract
//!
//! Operators (starting from highest priority):
//! * `!` - factorial (when used after a number or closing bracket)
//! * `-` - unary minus
//! * `**` - power
//! * `*`, `/`, `//` - multiplication, division, integer division
//! * `+`, `-` - addition, subtraction
//! * `&`, `^` - bitwise AND and XOR
//! * `|`, `<<`, `>>` - bitwise OR, SHL, and SHR
//! * `==`, `!=`, `>`, `<`, `>=`, and `<=` - comparison operators
//!
//! Predefined constants:
//! * `PI` - 3.14159...
//! * `E` - 2.71828...
//! * `PHI` - golden section - 1.6180...
//!
//! The calculator does not have special type for boolean numbers. When using logical
//! operators any zero value is treated as `false` and `true` otherwise. So, doing
//! double NOT transforms any number into big integer `1` or `0` depending on if the
//! original value was zero

#[macro_use]
extern crate pest_derive;

pub mod errors;
pub mod parse;
pub mod stack;
pub mod value;


src/errors.rs
use num_bigint::BigInt;
use num_rational::BigRational;
use std::fmt;

#[derive(PartialEq)]
pub enum CalcError {
    None,

    StrToFloat(String),
    StrToInt(String),
    StrToRatio(String),
    IntToFloat(BigInt),
    RatioToFloat(BigRational),
    AngleToFloat(String),
    FloatToInt(f64),
    FloatToRatio(f64),
    DividedByZero(String),
    InvalidShift(String),

    NotForComplex(String),
    NotForNegativeInt(String),

    OnlyInt(String),

    NotSupportedYet(String, String),
    InvalidAgrument(String, String),
    ArgumentOutOfRange(String, String, String),

    EmptyValue,
    InvalidOp(String),
    TooManyOps,
    OpenBracketMismatch,
    ClosingBracketMismatch,
    FunctionUnfinished(String),
    FunctionNoArgs(String),
    FunctionNotEnoughArgs(String, usize),
    EmptyExpression,
    InsufficientOps,
    VarUndeclared(String),

    NoRoots,
    ParseFailed(String),

    Unreachable,
}

impl fmt::Display for CalcError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match &self {
            CalcError::None => write!(f, "OK"),
            CalcError::StrToFloat(s) => write!(f, "Failed to convert '{}' to float", s),
            CalcError::StrToInt(s) => write!(f, "Failed to convert '{}' to integer", s),
            CalcError::StrToRatio(s) => write!(f, "Failed to convert '{}' to fraction", s),
            CalcError::AngleToFloat(s) => write!(f, "Failed to convert '{}' to angle", s),
            CalcError::IntToFloat(i) => write!(f, "Failed to convert integer {} to float", i),
            CalcError::RatioToFloat(r) => write!(f, "Failed to convert fraction {} to float", r),
            CalcError::FloatToInt(r) => write!(f, "Failed to convert float {} to integer", r),
            CalcError::FloatToRatio(r) => write!(f, "Failed to convert float {} to ratio", r),
            CalcError::DividedByZero(s) => write!(f, "'{}' divided by zero", s),
            CalcError::InvalidShift(s) => write!(f, "Cannot convert '{}' to usize", s),

            CalcError::NotForComplex(s) => write!(f, "Function '{}' is not supported for complex numbers", s),
            CalcError::NotForNegativeInt(s) => write!(f, "Function '{}' is not supported for negative integers", s),

            CalcError::OnlyInt(s) => write!(f, "{} supports only integers", s),

            CalcError::NotSupportedYet(func, typ) => {
                write!(f, "Function '{}' is not supported for {} numbers yet", func, typ)
            }
            CalcError::InvalidAgrument(func, val) => write!(f, "Invalid argument {} for function '{}'", func, val),
            CalcError::ArgumentOutOfRange(func, val, range) => {
                write!(f, "Argument {} of {} out of range({})", val, func, range)
            }

            CalcError::EmptyValue => write!(f, "Nor value neither operator found"),
            CalcError::InvalidOp(s) => write!(f, "Invalid operator '{}'", s),
            CalcError::TooManyOps => write!(f, "Too many operators"),
            CalcError::ClosingBracketMismatch => write!(f, "Mismatched closing bracket"),
            CalcError::OpenBracketMismatch => write!(f, "Mismatched opening bracket"),
            CalcError::FunctionUnfinished(s) => write!(f, "Closing bracket for function '{}' not found", s),
            CalcError::FunctionNoArgs(s) => write!(f, "Function '{}' requires an argument", s),
            CalcError::FunctionNotEnoughArgs(s, i) => write!(f, "Function '{}' requires at least {} arguments", s, i),
            CalcError::EmptyExpression => write!(f, "Nothing to calculate"),
            CalcError::InsufficientOps => write!(f, "Too many numbers"),

            CalcError::NoRoots => write!(f, "Equation has no roots"),
            CalcError::ParseFailed(s) => write!(f, "Failed to parse expression: {}", s),
            CalcError::VarUndeclared(s) => write!(f, "Variable '{}' not found", s),

            CalcError::Unreachable => write!(f, "unreachable"),
        }
    }
}

impl fmt::Debug for CalcError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match &self {
            CalcError::None => write!(f, "OK"),
            CalcError::StrToFloat(s) => write!(f, "Failed to convert '{}' to float", s),
            CalcError::StrToInt(s) => write!(f, "Failed to convert '{}' to integer", s),
            CalcError::StrToRatio(s) => write!(f, "Failed to convert '{}' to fraction", s),
            CalcError::AngleToFloat(s) => write!(f, "Failed to convert '{}' to angle", s),
            CalcError::IntToFloat(i) => write!(f, "Failed to convert integer {} to float", i),
            CalcError::RatioToFloat(r) => write!(f, "Failed to convert fraction {} to float", r),
            CalcError::FloatToInt(r) => write!(f, "Failed to convert float {} to integer", r),
            CalcError::FloatToRatio(r) => write!(f, "Failed to convert float {} to ratio", r),
            CalcError::DividedByZero(s) => write!(f, "'{}' divided by zero", s),
            CalcError::InvalidShift(s) => write!(f, "Cannot convert '{}' to usize", s),

            CalcError::NotForComplex(s) => write!(f, "Function '{}' is not supported for complex numbers", s),
            CalcError::NotForNegativeInt(s) => write!(f, "Function '{}' is not supported for negative integers", s),

            CalcError::OnlyInt(s) => write!(f, "{} supports only integers", s),

            CalcError::NotSupportedYet(func, typ) => {
                write!(f, "Function '{}' is not supported for {} numbers yet", func, typ)
            }
            CalcError::InvalidAgrument(func, val) => write!(f, "Invalid argument {} for function '{}'", func, val),
            CalcError::ArgumentOutOfRange(func, val, range) => {
                write!(f, "Argument {} of {} out of range({})", val, func, range)
            }

            CalcError::EmptyValue => write!(f, "Nor value neither operator found"),
            CalcError::InvalidOp(s) => write!(f, "Invalid operator '{}'", s),
            CalcError::TooManyOps => write!(f, "Too many operators"),
            CalcError::ClosingBracketMismatch => write!(f, "Mismatched closing bracket"),
            CalcError::OpenBracketMismatch => write!(f, "Mismatched opening bracket"),
            CalcError::FunctionUnfinished(s) => write!(f, "Closing bracket for function '{}' not found", s),
            CalcError::FunctionNoArgs(s) => write!(f, "Function '{}' requires an argument", s),
            CalcError::FunctionNotEnoughArgs(s, i) => write!(f, "Function '{}' requires at least {} arguments", s, i),
            CalcError::EmptyExpression => write!(f, "Nothing to calculate"),
            CalcError::InsufficientOps => write!(f, "Too many numbers"),

            CalcError::NoRoots => write!(f, "Equation has no roots"),
            CalcError::ParseFailed(s) => write!(f, "Failed to parse expression: {}", s),
            CalcError::VarUndeclared(s) => write!(f, "Variable '{}' not found", s),

            CalcError::Unreachable => write!(f, "unreachable"),
        }
    }
}


src/parse.rs
use pest::Parser;
use std::f64::consts::{E, PI};

use crate::errors::*;
use crate::stack::{Stack, FACTORIAL, PERCENT_ADD, PERCENT_DIV, PERCENT_MUL, PERCENT_SUB, STD_FUNCS, UNARY_MINUS};
use crate::value::*;

#[derive(Parser)]
#[grammar = "calc.pest"]
pub struct CalcParser;

pub const PHI: f64 = 1.618_033_988_749_895;
const LAST_RESULT: &str = "ans";

/// holds user user-defined variable
pub struct Variable {
    name: String,
    value: Value,
}
impl Variable {
    pub fn new(name: &str, value: Value) -> Self {
        Variable { name: name.to_string(), value }
    }
}

/// holds the current state of the engine, e.g list of all previously used variables
pub struct CalcState {
    variables: Vec<Variable>,
    is_last_value: bool,
    is_last_func: bool,
    pub has_alt: bool,
    pub alt_result: String,
}

impl Default for CalcState {
    fn default() -> CalcState {
        CalcState {
            variables: Vec::new(),
            is_last_value: false,
            is_last_func: false,
            has_alt: true,
            alt_result: "".to_owned(),
        }
    }
}

impl CalcState {
    pub fn new() -> Self {
        Default::default()
    }

    /// Returns a constant value by its name. Name is caseinsensitive
    pub fn constant(&self, name: &str) -> Option<Value> {
        let a = name.to_lowercase();
        match a.as_str() {
            "e" => Some(Value::Float(E)),
            "pi" => Some(Value::Float(PI)),
            "phi" | "golden" | "gold" => Some(Value::Float(PHI)),
            _ => None,
        }
    }

    /// Returns a variable value by its name. Name is caseinsensitive
    pub fn variable(&self, name: &str) -> Option<Value> {
        let low = name.to_lowercase();
        for v in &self.variables {
            if v.name == low {
                return Some(v.value.clone());
            }
        }
        None
    }

    /// Creates a new variable or replace existing one with a new value.
    /// The function does not check variable's name for validity. In case
    /// of its name conflicts with existing function name, the variable
    /// will be inaccessible from expression evaluator
    pub fn add_variable(&mut self, name: &str, val: Value) {
        let name = name.to_lowercase();
        for v in &mut self.variables {
            if v.name == name {
                v.value = val;
                return;
            }
        }
        self.variables.push(Variable::new(&name, val));
    }

    /// Returns the result of the last successful evaluation
    pub fn result(&self) -> Option<Value> {
        self.variable(LAST_RESULT)
    }

    /// Check if variable name is valid:
    /// - name convention
    /// - does not conflict with any constant
    /// - does not conflist with any function
    /// - does not conflict with special variables, like `ans`
    pub fn variable_name_validate(&self, name: &str) -> Result<(), &'static str> {
        let name = name.to_lowercase();
        if let Some(p) = name.find(|c: char| ('a'..='z').contains(&c)) {
            if p != 0 {
                return Err("Variable name must start with 'a'..'z'");
            }
        } else {
            return Err("Variable name must start with 'a'..'z'");
        }

        let p = name.find(|c: char| c != '_' && !('0'..='9').contains(&c) && !('a'..='z').contains(&c));
        if p.is_some() {
            return Err("Variable name must contain only Latin letters, digits, and underscore");
        }

        if self.constant(&name).is_some() {
            return Err("Cannot assign a new value to a constant");
        }
        if name == LAST_RESULT {
            return Err("The name is reserved for an internal variable");
        }

        for fname in STD_FUNCS.iter() {
            if fname == &name {
                return Err("Function name cannot be used as a variable");
            }
        }

        Ok(())
    }
}

macro_rules! process_value {
    ($id: ident, $stack: ident, $state: ident, $val: ident) => {
        if $state.is_last_func {
            $stack.push("(", None)?;
        } else if $state.is_last_value {
            $stack.push("*", None)?;
        }
        let v = Value::$id(&$val)?;
        $stack.push("", Some(v))?;
        if $state.is_last_func {
            $stack.push(")", None)?;
        }
        $state.is_last_value = true;
        $state.is_last_func = false;
    };
}

struct PrepRule {
    r: Rule,
    v: String,
}

fn fixup_last_prc_op(pairs: &mut Vec<PrepRule>) -> bool {
    let mut id: usize = pairs.len() - 1;
    let mut level = 0;
    while id > 0 {
        let r = pairs[id].r;
        let v = pairs[id].v.clone();
        match r {
            Rule::close_b => level += 1,
            Rule::open_b => level -= 1,
            Rule::operator if level == 0 => match v.as_str() {
                "+" => {
                    pairs[id].v = PERCENT_ADD.to_string();
                    return true;
                }
                "-" => {
                    pairs[id].v = PERCENT_SUB.to_string();
                    return true;
                }
                "*" => {
                    pairs[id].v = PERCENT_MUL.to_string();
                    return true;
                }
                "/" => {
                    pairs[id].v = PERCENT_DIV.to_string();
                    return true;
                }
                _ => return false,
            },
            _ => {}
        }
        id -= 1;
    }
    false
}

fn preprocess_expr(expr: &str) -> Result<Vec<PrepRule>, CalcError> {
    let pairs = match CalcParser::parse(Rule::expr, expr) {
        Ok(p) => p,
        // detailed error from pest parser
        // Err(e) => return Err(CalcError::ParseFailed(e.to_string())),
        // rcalc own error
        Err(..) => return Err(CalcError::ParseFailed("invalid expression".to_string())),
    };
    let mut is_last_prc = false;
    let mut preps: Vec<PrepRule> = Vec::new();
    for pair in pairs {
        let rule = pair.as_rule();
        let val = pair.as_span().as_str().to_lowercase();
        match rule {
            Rule::close_b | Rule::arg_sep | Rule::operator => {
                let is_prc = val == "%";
                if is_last_prc {
                    let _ = preps.pop();
                    if !fixup_last_prc_op(&mut preps) {
                        preps.push(PrepRule { r: Rule::operator, v: "%".to_string() });
                    }
                }
                preps.push(PrepRule { r: rule, v: val });
                is_last_prc = is_prc;
            }
            _ => {
                is_last_prc = val == "%";
                preps.push(PrepRule { r: rule, v: val });
            }
        }
    }
    if is_last_prc {
        let _ = preps.pop();
        if !fixup_last_prc_op(&mut preps) {
            preps.push(PrepRule { r: Rule::operator, v: "%".to_string() });
        };
    }
    Ok(preps)
}

/// evaluates a given expression and returns either result or error
pub fn eval(expr: &str, state: &mut CalcState) -> CalcResult {
    state.is_last_value = false;
    state.is_last_func = false;
    state.has_alt = false;

    let rules = preprocess_expr(expr)?;
    let mut stk = Stack::new();
    for pair in rules {
        let rule = pair.r;
        let val = pair.v;
        match rule {
            Rule::int | Rule::fulluint | Rule::hex | Rule::bin | Rule::oct => {
                process_value!(from_str_integer, stk, state, val);
            }
            Rule::float => {
                process_value!(from_str_float, stk, state, val);
            }
            Rule::ratio => {
                process_value!(from_str_ratio, stk, state, val);
            }
            Rule::degreefloat | Rule::fulldegree => {
                process_value!(from_str_angle, stk, state, val);
            }
            Rule::complex => {
                // distinguish between "1 - 2+i4" and "1 - -2+i4"
                #[allow(clippy::branches_sharing_code)]
                if state.is_last_value && val.starts_with('-') {
                    stk.push("-", None)?;
                    let slice = val[1..].to_string();
                    state.is_last_value = false;
                    state.is_last_func = false;
                    process_value!(from_str_complex, stk, state, slice);
                } else {
                    process_value!(from_str_complex, stk, state, val);
                }
            }
            Rule::open_b => {
                if state.is_last_value {
                    stk.push("*", None)?;
                }
                stk.push("(", None)?;
                state.is_last_value = false;
                state.is_last_func = false;
            }
            Rule::close_b => {
                stk.push(")", None)?;
                state.is_last_value = true;
                state.is_last_func = false;
            }
            Rule::arg_sep => {
                stk.push(";", None)?;
                state.is_last_value = false;
                state.is_last_func = false;
            }
            Rule::operator => {
                if val == "+" && !state.is_last_value {
                    state.is_last_value = false;
                } else if val == "-" && (!state.is_last_value || state.is_last_func) {
                    if state.is_last_func {
                        stk.push("(", None)?;
                        stk.push(")", None)?;
                        stk.push("-", None)?;
                    } else {
                        stk.push(UNARY_MINUS, None)?;
                    }
                    state.is_last_value = false;
                } else if val == "!" && state.is_last_value {
                    stk.push(FACTORIAL, None)?;
                    state.is_last_value = true;
                } else {
                    stk.push(&val, None)?;
                    state.is_last_value = false;
                }
                state.is_last_func = false;
            }
            Rule::ident => {
                if stk.is_func(&val) {
                    if state.is_last_value {
                        stk.push("*", None)?;
                    } else if state.is_last_func {
                        stk.increase_func_argc()?;
                    }
                    stk.push(&val, None)?;
                    state.is_last_value = false;
                    state.is_last_func = true;
                } else {
                    if let Some(v) = state.constant(&val) {
                        stk.push("", Some(v))?;
                    } else if let Some(v) = state.variable(&val) {
                        stk.push("", Some(v))?;
                    } else {
                        return Err(CalcError::VarUndeclared(val.to_string()));
                    }
                    state.is_last_value = true;
                    state.is_last_func = false;
                }
            }
            _ => return Err(CalcError::Unreachable),
        }
    }
    let output = stk.calculate();
    if let Ok(ref v) = output {
        state.add_variable(LAST_RESULT, v.clone());
        if stk.has_alt {
            state.has_alt = true;
            state.alt_result = stk.alt_result;
        }
    }
    output
}

#[cfg(test)]
mod tests {
    use super::*;
    use num_bigint::BigInt;
    use num_complex::Complex;
    use num_rational::BigRational;

    #[test]
    fn test_expr() {
        let mut state: CalcState = CalcState::new();
        let v = eval("2+3", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(5))));
        let v = eval("(3+2)(4-9)", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(-25))));
        let v = eval("sin(1;2;3)", &mut state);
        assert_eq!(v, Ok(Value::Float(1.0f64.sin())));
        let v = eval("(3+9)sin(1)", &mut state);
        assert_eq!(v, Ok(Value::Float(12.0f64 * 1.0f64.sin())));
        let v = eval("1\\2+3\\5", &mut state);
        assert_eq!(v, Ok(Value::Ratio(BigRational::new(BigInt::from(11), BigInt::from(10)))));
        let v = eval("1+2+i3", &mut state);
        assert_eq!(v, Ok(Value::Complex(Complex::new(3.0, 3.0))));
        let v = eval("1-2+i3", &mut state);
        assert_eq!(v, Ok(Value::Complex(Complex::new(-1.0, -3.0))));
        let v = eval("1--2+i3", &mut state);
        assert_eq!(v, Ok(Value::Complex(Complex::new(3.0, -3.0))));
        let v = eval("10+--5!/10", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(22))));
        let v = eval("10+-!!20*3", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(7))));
        let v = eval("1 + sin cos 2 * 10", &mut state);
        assert_eq!(v, Ok(Value::Float(2.0f64.cos().sin() * 10.0 + 1.0)));
        let v = eval("20 + 50 %", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(30))));
        let v = eval("40 + 30 - 50 %", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(55))));
        let v = eval("40 + 30 - 50 % - 10", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(45))));
        let v = eval("(40 + 30) - 50 %", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(35))));
        let v = eval("30 * 50 %", &mut state);
        assert_eq!(v, Ok(Value::Int(BigInt::from(15))));
        let v = eval("30 / 15 %", &mut state);
        assert_eq!(v, Ok(Value::Float(200.0)));
    }

    #[test]
    fn test_corner_cases() {
        let mut state: CalcState = CalcState::new();
        let exprs: [&'static str; 18] = [
            // auto: float -> complex
            "ln(-3) * sqrt(-5) + asin(5)",
            // exp is float, so the result is inaccurate f64
            "exp(100) * 12345678901234567890 * 987654321012345678901234567890",
            // exp is converted to big int, so the result is exact
            "round(exp(100)) * 12345678901234567890 * 987654321012345678901234567890",
            // a lot of single argument functions without brackets
            "sin cos sin cos sin cos 0.1",
            // should generate the same result as above
            "sin (cos( sin( cos (sin (cos (0.1))))))",
            // a lot of unary pluses
            "2+++++3sqrt 5",
            // double factorial should have higher priority than multiplication
            "3!!*2",
            // negated complex with positive real part
            "1.0+9.2i-2.3-4.8i",
            // negated complex with negative real part
            "1.0+9.2i--2.3-4.8i",
            // raise to negative floating point degree
            "2**-e",
            // different cases to test automatic insertion of brackets and multiplication sign
            "sin cos 2 30",
            "sin cos (2*30)",
            "sin (cos (2))*30",
            "sin (cos 2)*30",
            "sin cos 2*30",
            "sin(1)sin(1)",
            "sin 1 sin 1",
            "sin 1 sin(1)",
        ];
        let ress: [&'static str; 18] = [
            "-5.45401840424583+0.16414008881733767i",
            "3.2776919587749234e92",
            "327769195877492364977545717691598341790414097481269687584256420446284592691174142251735449600",
            "0.7270679494248203",
            "0.7270679494248203",
            "8.70820393249937",
            "1440",
            "-1.2999999999999999+14.0i",
            "3.3+14.0i",
            "0.15195522325791298",
            "-12.127174615567974",
            "-0.8148167252856553",
            "-12.127174615567974",
            "-12.127174615567974",
            "-12.127174615567974",
            "0.7080734182735712",
            "0.7080734182735712",
            "0.7080734182735712",
        ];

        for (i, expr) in exprs.to_vec().iter().enumerate() {
            let res = eval(expr, &mut state);
            let res_str = match res {
                Ok(v) => format!("{}", v),
                Err(e) => format!("{:?}", e),
            };
            assert_eq!(res_str, ress[i].to_string());
        }
    }
}
